### PR TITLE
[BEAM-3415] Amend TestPipeline to work with JUnit5

### PR DESCRIPTION
**Pull request by emilssolmanis**
This makes `TestPipeline` implement a JUnit5 extension, making it usable with the JUnit5 extensions mechanism (what used to be "Rules" in JUnit4). 

It's semantically the same code, the only outward noticeable change would be JUnit5 uses `@Tag` instead of `@Category`, and they're string values not classes, so `@Category(NeedsRunner.class)` becomes `@Tag("needsRunner")` -- open to guidance on the value here, happy to consider the existing one a placeholder.

This also adds a separate JUnit5 test for the capablities of `TestPipeline` and a JUnit5 run in the `build.gradle`. It should be fine running side-by-side (both in theory and at least my experience) since they use different annotations etc., JUnit5 will only pick up the Jupiter annotations and JUnit4 will only use the `org.junit` ones.

R: @lukecwik 
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by emilssolmanis** [comment]
Implementing JUnit5 hooks in the existing `TestPipeline` ended up breaking compilation across tons of components, due to the now required Jupiter API, which I figured would be unacceptable for clients.

I ended up opening up `TestPipeline` a bit (setting the constructor to `protected`, and some select fields), and extending it with a separate `TestPipelineJunit5` class. This way people should hopefully not get compile problems unless they use the new class.

Would love to hear thoughts and iterate though, if this is not acceptable. Also happy to just collapse them back into `TestPipeline` and add the dependency to all the modules, if we think that level of breaking everyone's stuff is acceptable.

**Comment by emilssolmanis** [comment]
Also just noticed template says comment, not tag in description, so just in case there's behind-the-scenes automation

R: @lukecwik

**Comment by boyuanzz** [comment]
@tysonjh or @kennknowles 
Could you please help review this PR or find someone help this?

**Comment by TheNeuralBit** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by emilssolmanis** [comment]
@boyuanzz @tysonjh @kennknowles @TheNeuralBit any feedback / changes needed? How would we go about getting this merged?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java [hunk 1]
```diff
@@ -107,7 +107,7 @@
 
   private final PipelineOptions options;
 
-  private static class PipelineRunEnforcement {
+  protected static class PipelineRunEnforcement {
```
Comment (TheNeuralBit) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java:110]: Could these be package-private instead?
```suggestion
  static class PipelineRunEnforcement {
```

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipelineJunit5.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.junit.jupiter.api.extension.AfterEachCallback;
+import org.junit.jupiter.api.extension.BeforeEachCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+
+/**
+ * An extension of {@link TestPipeline} that implements the JUnit5 Extensions model.
+ *
+ * <p>Usage is the same as {@link TestPipeline}, but the instantiation has to be done via JUnit5's
+ * {@link org.junit.jupiter.api.extension.RegisterExtension} annotation mechanism, like
```
Comment (TheNeuralBit) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipelineJunit5.java:33]: It looks like you're missing an example here

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipelineJunit5.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+ *  }
+ * </code></pre>
+ *
+ * <p>See also the <a href="https://beam.apache.org/contribute/testing/">Testing</a> documentation
+ * section.
+ */
+@SuppressWarnings({
+  "nullness" // TODO(https://issues.apache.org/jira/browse/BEAM-10402)
+})
+public class TestPipelineJunit5 extends TestPipeline
```
Comment (TheNeuralBit) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipelineJunit5.java:57]: Is there any way we could provide a single `TestPipeline` that can be used with both Junit5 and Junit4? I think that would be preferable.

**Review by TheNeuralBit — COMMENTED** [review]
Sorry I'm not very familiar with the `TestPipeline` implementation, but I can help review.  I have a few initial comments/questions

**Comment by kennknowles** [comment]
Haven't had time to look deeply, but just checking that this will still work with JUnit 4?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipelineJunit5.java (hunk 2 above)
Comment (emilssolmanis) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipelineJunit5.java:33]: Will add.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java (hunk 1 above)
Comment (emilssolmanis) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java:110]: They can be, it could be argued that `protected` allows for some extensibility, i.e., if they were protected to begin with, I could have added a subclass implementing the 2 JUnit5 interfaces somewhere in our own test code instead of making a whole Beam PR.

I don't know if there's an **actual** use case mind you, I'm unfamiliar with TestNG and other test frameworks and whether there's any need, so perfectly happy to make them package-private to hide from clients.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipelineJunit5.java (hunk 3 above)
Comment (emilssolmanis) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipelineJunit5.java:57]: @TheNeuralBit not without breaking clients, unfortunately. That was my first idea, but adding the JUnit5 imports to `TestPipeline` then makes JUnit5 a compile dependency for anything that uses `TestPipeline`.

I don't mind adding the dependency to the tens of Beam modules, but the more important part is of course that it would force anyone doing testing with JUnit4 to add a JUnit5 API dependency as well just for their code to compile.

Currently the testing deps are all `provided`, the alternative is marking some of them as `compile`, but I'm not sure that's a good approach either.

**Comment by emilssolmanis** [comment]
@kennknowles yes, all the JUnit5 stuff is additive and just enables support, nothing changes with respect to JUnit4.

**Comment by emilssolmanis** [comment]
@TheNeuralBit yeah I saw that, and didn't understand why they'd done it. I'd be keen to hear from the author themselves @rmannibucau if you have context, but my guess would be both JUnit5 and Beam have evolved in the 2 years since. Purely just implementing the before / after `TestExecutionCAllback` interfaces from JUnit5 seems to work, the added JUnit5 test suite verifies the same things the JUnit4 ones does, this includes the various calls to `PAssert`. Anecdotally I also did some local testing on a company project where we're looking to utilise this by copy-pasting all the relevant bits and just implementing the 2 JUnit5 extensions, and all our tests seem to pass by just swapping out the annotations and runner.

tl dr there's a test suite exercising `PAssert` calls and local testing would seem to confirm that the code is fine.

I don't mind an extension, since that solves the whole "clients would be required to depend on both versions of JUnit" debacle, but not sure that's justified for 2 classes (the TestPipeline and the sys prop restorer util). If that's the preferred route I'd love some thoughts on whether we want to just copy-paste the existing TestPipeline and allow for potential divergence of the JUnit4 / 5 implementations, or whether the current inheritance coupling seems appropriate.

**Comment by rmannibucau** [comment]
@emilssolmanis original migration goal was to support natively junit5 (and not rely on junit4 imports which has several pitfalls in integrations layers (thinking to maven/gradle plugins and IDE) + in user land with completion packages - "why my @test does not run"...wrong package). At that time beam community kind of rejected this as not being a goal/needed for google - at that time - so got rid of this need by providing an alternate testing solution and bypassing beam native testing stack which was not moving to user direction fast enough. Was still the case a few months ago but maybe it changed in between (I don't use beam at all anymore so stopped to follow the project).

**Comment by kennknowles** [comment]
Yea, it is a tough problem. It doesn't really have to do with Google or any specific company.

Judging by any data I can find, JUnit4 is still by far the dominant version in use. Why? JUnit5 isn't so much an upgrade but just a totally different project with a shared name and some similar concepts, and also with API/functionality regressions that are not a priority for JUnit team. The migration cost for a large organization is huge, as much as migrating to an unrelated test framework, and benefits not enough to justify the investment. It reminds me of Python 2 to 3.

This is just my basic understanding of the current status.

Why does Beam need to care about what test frameworks others are using? There are real reasons:

 - We export our "Beam Model Compliance" test suite via JUnit4 category. So it is a critical breaking change to do that differently.
 - So we also do want to support people running tests in their own infrastructure to test their runners.
 - And of course, being part of the dominant version gives network effects of the best maintenance and support via tooling.

Of course, in the same spirit, if there are Beam users/contributors who want to run our tests in their infrastructure that uses JUnit5 it is nice to support them. So I very much support figuring out a way to help both to work.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java (hunk 1 above)
Comment (kennknowles) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java:110]: Noting that you can always extend, in probably a better way, via object composition. You can subclass `Pipeline` and forward method calls to a local `TestPipeline` variable.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipelineJunit5.java (hunk 3 above)
Comment (kennknowles) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipelineJunit5.java:57]: Refining this question: what we really want to support is organizations/individuals _running_ tests with JUnit5. Authoring the tests probably needs to stay JUnit4, for reasons in my other comment. But we want to be compatible with custom test infrastructure that someone might be using in their own project.

Comment (kennknowles) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipelineJunit5.java:57]: Correction: we don't want _Beam_ tests to be written in JUnit5 style. Allowing Beam users to author in their preferred style is great.

**Comment by rmannibucau** [comment]
> Judging by any data I can find, JUnit4 is still by far the dominant version in use. Why? JUnit5 isn't so much an upgrade but just a totally different project with a shared name and some similar concepts, and also with API/functionality regressions that are not a priority for JUnit team. The migration cost for a large organization is huge, as much as migrating to an unrelated test framework, and benefits not enough to justify the investment. It reminds me of Python 2 to 3.

Using CI data yes because it exists but using new project stacks not really because projects start on up to date framework and junit4 said they will not upgrade the framework (until a CVE pops up) so Junit4 can be seen as dead as of today and JUnit5 the main alternative. I know it is hard to have the difference in terms of data but people doing service on new projects will surely confirm what i saw over the last year.
In terms of cost, it is generally cheap to have done it on big suites, the only cost is about migrating rules to extensions and it is not that costly in practise and rest is mainly a sed so not as bad as python even if it can look like that upfront.
Last point is that with junit 5.7, junit5 caught back its lateness on junit4 (mainly rules which were enabling more before) so guess it is time for beam to move anyway. It can't go back anyway :).

> We export our "Beam Model Compliance" test suite via JUnit4 category. So it is a critical breaking change to do that differently.

JUnit 5 has groups and it does not have to break, guess junit4 should stay in a beam-junit4 module and jupiter move to a beam-junit5 one. Current beam-core ("put it all") will become a kind of bom (with an empty jar to not break deps too much?) and core can move in a beam-core-engine, no breaking change for anyone if desired.

To have done the work, there is no real reason to be that coupled with a testing framework - don't get me wrong, it was natural in a first version of the framework, I'm not blaming - so I guess it is mainly about abstracting the few basic primitives (primitives not the whole hamcrest stuff ;)).

Side note: it can be an opportunity to split beam internal harnessing and user testing API. Using direct runner enable to cover user case quite well and does not need the test pipeline or so which mainly brings constraints and as much off toggle than feature so rethinking the API and maybe splitting both makes sense to me.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipelineJunit5.java (hunk 3 above)
Comment (emilssolmanis) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipelineJunit5.java:57]: @kennknowles agree with all of that. The code here _in my opinion_ does that, the step further here would be splitting out the testing modules so it doesn't all live in `-core`. So there'd be a `beam-junit4` and a `beam-junit5`, or alternatively junit4 remains in `core` for backwards compatibility and the Junit5 code is a new module. Happy to do either of those options if we think that's warranted, I'm unfamiliar with how Beam decides on what warrants a new module.

**Comment by emilssolmanis** [comment]
@kennknowles @rmannibucau 
sorry for the lack of response, been a bit snowed under at work.

So 2 points I can extract from the conversation so far:
1. the dominant version of Junit and all that stems from there -- I think this would require data, and I'm not sure we have that available. I'm with @rmannibucau in that most new projects will start in Jupiter, and most old projects stay in Junit4 purely by inertia. I don't think there's a right or wrong solution, and supporting either seems simple enough, at least from what I've seen so far, so being independent of the testing framework sounds like a good thing to aim for and should be achievable.
2. > We export our "Beam Model Compliance" test suite via JUnit4 category. So it is a critical breaking change to do that differently.

Not sure I follow, is this documented somewhere I can have a read about? Mainly I don't immediately understand how this would affect clients.

**Comment by rmannibucau** [comment]
@emilssolmanis I assume @kennknowles speaks about https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/testing/ValidatesRunner.java and the related tests which validates a runner impl complies to the spec (if about app code then it is not a blocker at all since it will fall in junit5 extension impl). In this spec case the runner will have to run the tests through junit4 which is not a big deal for now (not all runners - outside the project itself -try to pass these tests and for now it is a bean internal thing anyway). So overall no blocker to support a junit5 testing solution IMHO.

**Comment by MaximeFrancoeur** [comment]
what is the progress on this PR? I hope we can merge quickly :)

**Comment by aromanenko-dev** [comment]
@MaximeFrancoeur Seems like work is suspended because of different opinions on this.
Imho, this sensitive topic requires a discussion on mailing lists (dev@ and user@) and, probably, a vote before taking a final decision.

**Comment by BStilesHSQ** [comment]
Hope discussions on that are continuing. Would like to fully migrate my projects that use Beam away from JUnit 4.

**Comment by kennknowles** [comment]
I am not aware of a discussion going on `dev@beam.apache.org`. You could definitely start it up!
